src/Microsoft.ML.Data/Transforms/SelectColumnsTransform.cs

```diff
@@ -0,0 +1,658 @@
+                                                      string.Join(",", invalidColumns));
+
+            return new Mapper(this, inputSchema).Schema;
+        }
+
+        public IRowToRowMapper GetRowToRowMapper(Schema inputSchema)
+        {
+            _host.CheckValue(inputSchema, nameof(inputSchema));
+            if (!IgnoreMismatch &&
+                !IsSchemaValid(inputSchema.GetColumns().Select(x=>x.column.Name),
+                                SelectColumns,
+                                out IEnumerable<string> invalidColumns))
+                  throw _host.ExceptSchemaMismatch(nameof(inputSchema), "input",
+                                                      string.Join(",", invalidColumns));
+
+            return new SelectColumnsDataTransform(_host, this,
+                                                  new Mapper(this, inputSchema),
+                                                  new EmptyDataView(_host, inputSchema));
+        }
+
+        public IDataView Transform(IDataView input)
+        {
+            _host.CheckValue(input, nameof(input));
+            if (!IgnoreMismatch &&
+                !IsSchemaValid(input.Schema.GetColumns().Select(x=>x.column.Name),
+                                SelectColumns,
+                                out IEnumerable<string> invalidColumns))
+                  throw _host.ExceptSchemaMismatch(nameof(input), "input",
+                                                      string.Join(",", invalidColumns));
+
+            return new SelectColumnsDataTransform(_host, this, new Mapper(this, input.Schema), input);
+        }
+
+        private sealed class Mapper
+        {
+            private readonly IHost _host;
+            private readonly Schema _inputSchema;
+            private readonly int[] _outputToInputMap;
+
+            public ISchema InputSchema => _inputSchema;
+
+            public Schema Schema { get; }
+
+            public Mapper(SelectColumnsTransform transform, ISchema inputSchema)
+            {
+                _host = transform._host.Register(nameof(Mapper));
+                _inputSchema = Runtime.Data.Schema.Create(inputSchema);
+
+                _outputToInputMap = BuildOutputToInputMap(transform.SelectColumns,
+                                                            transform.KeepColumns,
+                                                            transform.KeepHidden,
+                                                            _inputSchema);
+                Schema = GenerateOutputSchema(_outputToInputMap, _inputSchema);
+            }
+
+            public int GetInputIndex(int outputIndex)
+            {
+                _host.Assert(0 <= outputIndex && outputIndex < _outputToInputMap.Length);
+                return _outputToInputMap[outputIndex];
+            }
+
+            private static int[] BuildOutputToInputMap(IEnumerable<string> selectedColumns,
+                                                         bool keepColumns,
```

Comment:
>bool [](start = 57, length = 4)

this is a non-standard padding. Standard is 4 spaces to the right of 'private'

Reply:
Does this apply to function calls as well? I usually try to keep the parameters together for readability. 

---
In reply to: [226763760](https://github.com/dotnet/machinelearning/pull/1269#discussion_r226763760) [](ancestors = 226763760)